internal/services/pam/pam_test.go

```diff
@@ -0,0 +1,596 @@
+
+	gotResp, _ = testClient.GetPreviousBroker(context.Background(), &authd.GPBRequest{Username: "does not have a broker"})
+	require.Empty(t, gotResp.GetPreviousBroker(), "GetPreviousBroker should return nil when the user has no broker assigned")
+}
+
+func TestSelectBroker(t *testing.T) {
+	t.Parallel()
+
+	tests := map[string]struct {
+		brokerID string
+		username string
+
+		wantErr bool
+	}{
+		"Successfully select a broker and creates the session": {username: "success"},
+
+		"Error when username is empty":                    {wantErr: true},
+		"Error when brokerID is empty":                    {username: "empty broker", brokerID: "-", wantErr: true},
+		"Error when broker does not exist":                {username: "no broker", brokerID: "does not exist", wantErr: true},
+		"Error when broker does not provide a session ID": {username: "NS_no_id", wantErr: true},
+		"Error when starting the session":                 {username: "NS_error", wantErr: true},
+	}
+	for name, tc := range tests {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			if tc.brokerID == "" {
+				tc.brokerID = generatedID
+			} else if tc.brokerID == "-" {
+				tc.brokerID = ""
+			}
+
+			sbRequest := &authd.SBRequest{
+				BrokerId: tc.brokerID,
+				Username: tc.username,
+			}
+			sbResp, err := testClient.SelectBroker(context.Background(), sbRequest)
+			if tc.wantErr {
+				require.Error(t, err, "SelectBroker should return an error, but did not")
+				return
+			}
+			require.NoError(t, err, "SelectBroker should not return an error, but did")
+			got := fmt.Sprintf("ID: %s\nEncryption Key: %s\n",
+				strings.ReplaceAll(sbResp.GetSessionId(), tc.brokerID, "BROKER_ID"),
+				sbResp.GetEncryptionKey())
+			want := testutils.LoadWithUpdateFromGolden(t, got)
+			require.Equal(t, want, got, "SelectBroker returned an unexpected response")
+		})
+	}
+}
+
+func TestGetAuthenticationModes(t *testing.T) {
+	t.Parallel()
+
+	tests := map[string]struct {
+		brokerID           string
+		sessionID          string
+		username           string
+		supportedUILayouts []*authd.UILayout
+
+		wantErr bool
+	}{
+		"Successfully get authentication modes":          {},
+		"Successfully get multiple authentication modes": {username: "GAM_multiple_modes"},
+
+		"Error when sessionID is empty":           {sessionID: "-", wantErr: true},
+		"Error when passing invalid layout":       {supportedUILayouts: []*authd.UILayout{emptyType}, wantErr: true},
+		"Error when broker does not exist":        {sessionID: "no broker", brokerID: "does not exist", wantErr: true},
+		"Error when getting authentication modes": {username: "GAM_error", wantErr: true},
+		"Error when broker returns invalid modes": {username: "GAM_invalid", wantErr: true},
+	}
+	for name, tc := range tests {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			if tc.brokerID == "" {
+				tc.brokerID = generatedID
+			}
+
+			if tc.brokerID != "does not exist" {
+				if tc.username == "" {
```

Comment:
I seems that we have quite some duplication over the Selection, returning the session ID. Should that be set in a testhelper in that file?